Issue 30525

francis-pouatcha opened the pull request:
- Closes #30525 
- Removed mandatory nbf from credential as per spec.
- Changed 'code' access token request key, to 'pre-authorized_code' for the correponding grant type

francis-pouatcha committed:
Enhance SupportedCredentialConfiguration to support optional claims object as defined in OpenID for Verifiable Credential Issuance specification.

Signed-off-by: Francis Pouatcha <francis.pouatcha@adorsys.com>

francis-pouatcha committed:
Merge branch 'keycloak:main' into issue-30419

francis-pouatcha committed:
Modifications from issue comment https://github.com/keycloak/keycloak/issues/30419#issuecomment-2169408107

vct claim is mandatory for sd-jwt. Can be ommited by other formats
display shall be an array
display object can be configured using an escaped json, as dot notation will be cumbersome with arrays
proof_types_supported can be configured using an escaped json, as each proof type has different configuration parameters.

Signed-off-by: Francis Pouatcha <francis.pouatcha@adorsys.com>

francis-pouatcha committed:
Fix on issue 30525

Signed-off-by: Francis Pouatcha <francis.pouatcha@adorsys.com>

francis-pouatcha committed:
Replaced code param with pre-authorized_code as stated by spec

Signed-off-by: Francis Pouatcha <francis.pouatcha@adorsys.com>

francis-pouatcha committed:
Removed mandatory presence of nbf claim as it is optional in the spec

Signed-off-by: Francis Pouatcha <francis.pouatcha@adorsys.com>

francis-pouatcha committed:
Serializing a proof type enum

Signed-off-by: Francis Pouatcha <francis.pouatcha@adorsys.com>

francis-pouatcha committed:
Fix: verifier wallet provided proof

Signed-off-by: Francis Pouatcha <francis.pouatcha@adorsys.com>

francis-pouatcha committed:
Merge branch 'keycloak:main' into issue-30525

francis-pouatcha committed:
Merged from main

Signed-off-by: Francis Pouatcha <francis.pouatcha@adorsys.com>

francis-pouatcha committed:
Fix: removed dupplicate instantiation of service

Signed-off-by: Francis Pouatcha <francis.pouatcha@adorsys.com>

francis-pouatcha committed:
Fix(test): set code request param grant ty urn:ietf:params:oauth:grant-type:pre-authorized_code to pre-authorized_code instead of code as per spec.

Signed-off-by: Francis Pouatcha <francis.pouatcha@adorsys.com>

francis-pouatcha committed:
Fix(test): set code request param grant ty urn:ietf:params:oauth:grant-type:pre-authorized_code to pre-authorized_code instead of code as per spec.

Signed-off-by: Francis Pouatcha <francis.pouatcha@adorsys.com>

francis-pouatcha committed:
Fix(test): remove check of nbf from tests, as nbf claim not mandatory as per spec.

Signed-off-by: Francis Pouatcha <francis.pouatcha@adorsys.com>

mposolda commented:
@tnorimat @wistefan Anyone available to review this PR please?

tnorimat commented:
@mposolda I see, I will review the PR, but it takes a lot of time to complete the review.

tnorimat commented:
@mposolda also, it might be better if @wistefan could review the PR.

wistefan commented:
@mposolda @tnorimat I'll review before end of tomorrow

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/VCSigningServiceProviderFactory.java:
```diff
@@ -44,13 +44,15 @@ public interface VCSigningServiceProviderFactory extends ComponentFactory<Verifi
 
     public static ProviderConfigurationBuilder configurationBuilder() {
         return ProviderConfigurationBuilder.create()
-                .property(SigningProperties.KEY_ID.asConfigProperty());
```
If removed here, it should be added to the individual providers for Jwt, SdJwt or LD. Why not leaving it here, but just removing the required-constraint in l53?

francis-pouatcha replied:
Resolved with [39100ed](https://github.com/keycloak/keycloak/commit/39100edc1281422399b045012acb3cb27062fbf6)

wistefan reviewed (changes requested):
Looks good in general, just some smaller things.

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/SdJwtSigningService.java:
```diff
@@ -141,4 +178,164 @@ public String signCredential(VerifiableCredential verifiableCredential) {
+                        "Proof not produced for this audience. Audience claim must be: " + credentialIssuer + " but are " + Arrays.asList(proofPayload.getAudience())));
+
+        // Validate mandatory iat.
+        // I do not understand the rationale behind requiring a issue time if we are not checking expiration.
```
While the specification suggests checking `iat`, the absence of a defined "acceptable time window" introduces potential for incompatibility. If the goal is to ensure proof actuality, I propose relying on a server-provided `c_nonce`. This allows the server to directly track proof validity, mitigating the clock synchronization issues that arise when verifying client-generated timestamps.

francis-pouatcha replied:
I mean the only reasonable exit i see is to check the `iat` against the `exp` of the bearer token. But we are still dealing with cloks from different worlds. There is still rooms for false positives.

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/SdJwtSigningService.java:
```diff
@@ -141,4 +178,164 @@ public String signCredential(VerifiableCredential verifiableCredential) {
+    private void validateProofPayload(VCIssuanceContext vcIssuanceContext, AccessToken proofPayload) throws VCIssuerException {
+        // azp is the id of the client, as mentioned in the access token used to request the credential.
+        // Token provided from user is obtained with a clientId that support the oidc login protocol.
+        // oid4vci client doesn't. But it is the client needed at the credential endpoint.
```
Ok.

francis-pouatcha replied:
Resolved with [39100ed](https://github.com/keycloak/keycloak/commit/39100edc1281422399b045012acb3cb27062fbf6)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/SdJwtSigningService.java:
```diff
@@ -118,18 +154,19 @@ public String signCredential(VerifiableCredential verifiableCredential) {
         rootNode.put(CREDENTIAL_ID_CLAIM, JwtSigningService.createCredentialId(verifiableCredential));
 
+        // add the key binding if any
+        if(jwk!=null){
```
Thanks.

francis-pouatcha replied:
Resolved with [39100ed](https://github.com/keycloak/keycloak/commit/39100edc1281422399b045012acb3cb27062fbf6)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/SdJwtSigningService.java:
```diff
@@ -141,4 +178,164 @@ public String signCredential(VerifiableCredential verifiableCredential) {
+    private void checkCryptographicKeyBinding(VCIssuanceContext vcIssuanceContext){
+        // Make sure we are dealing with a jwk proof.
+        if (vcIssuanceContext.getCredentialConfig().getCryptographicBindingMethodsSupported() == null ||
+                !vcIssuanceContext.getCredentialConfig().getCryptographicBindingMethodsSupported().contains("jwk")) {
```
Thanks

francis-pouatcha replied:
Resolved with [39100ed](https://github.com/keycloak/keycloak/commit/39100edc1281422399b045012acb3cb27062fbf6)

IngridPuppet commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java:
As a consequence of expecting a `CredentialRequest` object, the `requestCredential` method makes it mandatory for consumers to submit both a `format` and a `credential_identifier` field. However, according to the specs, these fields MUST NOT be expected at the same time. See https://openid.github.io/OpenID4VCI/openid-4-verifiable-credential-issuance-wg-draft.html#name-credential-request

Moreover, for the Pre-Authorized Code flow specifically, the chain of requirements in the specs appears not to permit requesting a credential by `credential_identifier`. Did you mean to enforce `format` and `vct` instead?

francis-pouatcha replied:
@wistefan, we're facing an issue here. We're currently using Format + credential_identifier to look up the signing service. To comply with the specification, we need to resolve the signing service based on:

* **Either** the credential_identifier alone, if available,
* **Or** the Format + protocol-specific identifier (e.g., vct).

francis-pouatcha replied:
Resolved with [39100ed](https://github.com/keycloak/keycloak/commit/39100edc1281422399b045012acb3cb27062fbf6)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/SdJwtSigningService.java:
```diff
@@ -141,4 +178,164 @@ public String signCredential(VerifiableCredential verifiableCredential) {
+     * @throws IllegalStateException: is credential type badly configured
+     * @throws IOException
+     */
+    private JWK validateProof(VCIssuanceContext vcIssuanceContext) throws VCIssuerException, JWSInputException, VerificationException, IOException {
```
Added JwtProofBasedSigningService but only consumed by SdJwtSigningService for the moment. We still need to find the spec on how to consume a proof in JwtSigningService.

francis-pouatcha replied:
Resolved with [39100ed](https://github.com/keycloak/keycloak/commit/39100edc1281422399b045012acb3cb27062fbf6)

tnorimat commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java:
```diff
@@ -423,19 +433,20 @@ private Response getErrorResponse(ErrorType errorType) {
-        if (Optional.ofNullable(vcType).filter(type -> !type.isEmpty()).isEmpty()) {
-            throw new BadRequestException("No VerifiableCredential-Type was provided in the request.");
+        if (Optional.ofNullable(vcScope).filter(scope -> !scope.isEmpty()).isEmpty()) {
+            throw new BadRequestException("No VerifiableCredential-Scope was provided in the request.");
```
I am afraid that it causes 500 instead of 401. Could you test this point by writing an integration test and run it with `-Pauth-server-quarkus` ?

francis-pouatcha replied:
Resolved with [346b41](https://github.com/keycloak/keycloak/commit/346b4151539737e83126f8cc9705b462441c1081)

tnorimat commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java:
```diff
@@ -392,11 +392,21 @@ private Object getCredential(UserSessionModel userSessionModel, String vcType, F
+
+        return Optional.ofNullable(signingService)
+                .map(service -> service.signCredential(vcIssuanceContext))
+                .orElseThrow(() -> new IllegalArgumentException(
```
If this exception is thrown, which kind of an error response a holder receive?

francis-pouatcha replied:
Resolved with [346b41](https://github.com/keycloak/keycloak/commit/346b4151539737e83126f8cc9705b462441c1081)

tnorimat commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/SigningService.java:
```diff
@@ -34,18 +41,64 @@ public abstract class SigningService<T> implements VerifiableCredentialsSigningS
     protected KeyWrapper getKey(String kid, String algorithm) {
+        // Allow the service to work with the active key if keyId is null
+        // And we still have to figure out how to proceed with key rotation
+        if(keyId==null){
```
It might be better to be `if (keyId == null) {`

francis-pouatcha replied:
Resolved with [346b41](https://github.com/keycloak/keycloak/commit/346b4151539737e83126f8cc9705b462441c1081)

tnorimat reviewed (changes requested):
@francis-pouatcha Hello, I addred review comments. Could you check them?

tnorimat commented:
@francis-pouatcha Hello, IMO, the issue (https://github.com/keycloak/keycloak/issues/30525) includes five points, so it might be helpful for reviewers to send a PR per each point separately, which reduces burdens for reviewing a PR.

francis-pouatcha committed:
Processing feedbacks from first review comments

Signed-off-by: Francis Pouatcha <francis.pouatcha@adorsys.com>

francis-pouatcha committed:
Processing feedbacks from first review comments

Signed-off-by: Francis Pouatcha <francis.pouatcha@adorsys.com>

francis-pouatcha committed:
Merging from main

Signed-off-by: Francis Pouatcha <francis.pouatcha@adorsys.com>

francis-pouatcha committed:
Processing review comments

Signed-off-by: Francis Pouatcha <francis.pouatcha@adorsys.com>

francis-pouatcha committed:
Deleting from last main merge

Signed-off-by: Francis Pouatcha <francis.pouatcha@adorsys.com>

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/VerifiableCredentialsSigningService.java:
```diff
@@ -17,6 +17,9 @@
+import org.keycloak.protocol.oid4vc.issuance.VCIssuanceContext;
+import org.keycloak.protocol.oid4vc.issuance.VCIssuerException;
+import org.keycloak.protocol.oid4vc.model.Format;
 import org.keycloak.protocol.oid4vc.model.VerifiableCredential;
```
Resolved with [346b41](https://github.com/keycloak/keycloak/commit/346b4151539737e83126f8cc9705b462441c1081)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/SigningService.java:
```diff
@@ -34,18 +41,64 @@ public abstract class SigningService<T> implements VerifiableCredentialsSigningS
+        return signatureProvider.verifier(getKeyWraper(jwk, jwsAlgorithm, KeyUse.SIG));
+    }
+
+    private KeyWrapper getKeyWraper(JWK jwk, String algorithm, KeyUse keyUse) {
```
Resolved with [346b41](https://github.com/keycloak/keycloak/commit/346b4151539737e83126f8cc9705b462441c1081)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/SigningService.java:
```diff
@@ -34,18 +41,64 @@ public abstract class SigningService<T> implements VerifiableCredentialsSigningS
     protected KeyWrapper getKey(String kid, String algorithm) {
+        // Allow the service to work with the active key if keyId is null
+        // And we still have to figure out how to proceed with key rotation
+        if(keyId==null){
```
Resolved with [346b41](https://github.com/keycloak/keycloak/commit/346b4151539737e83126f8cc9705b462441c1081)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/SdJwtSigningService.java:
```diff
@@ -141,4 +178,164 @@ public String signCredential(VerifiableCredential verifiableCredential) {
\ No newline at end of file
+    @Override
+    public String locator() {
+        return VerifiableCredentialsSigningService.locator(format,vcConfigId);
```
Resolved with [346b41](https://github.com/keycloak/keycloak/commit/346b4151539737e83126f8cc9705b462441c1081)

francis-pouatcha commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/issuance/signing/OID4VCTest.java:
```diff
@@ -105,7 +106,11 @@ public static KeyWrapper getECKey(String keyId) {
             kw.setPublicKey(keyPair.getPublic());
             kw.setUse(KeyUse.SIG);
-            kw.setKid(keyId);
+            if(keyId!=null) {
```
Resolved with [346b41](https://github.com/keycloak/keycloak/commit/346b4151539737e83126f8cc9705b462441c1081)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/SdJwtSigningService.java:
```diff
@@ -53,6 +71,9 @@ public class SdJwtSigningService extends SigningService<String> {
     private static final String NOT_BEFORE_CLAIM ="nbf";
```
Resolved with [346b41](https://github.com/keycloak/keycloak/commit/346b4151539737e83126f8cc9705b462441c1081)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/SigningProperties.java:
```diff
@@ -29,13 +29,16 @@ public enum SigningProperties {
 
     ISSUER_DID("issuerDid", "Did of the issuer.", "Provide the DID of the issuer. Needs to match the provided key material.", ProviderConfigProperty.STRING_TYPE, null),
     KEY_ID("keyId", "Id of the signing key.", "The id of the key to be used for signing credentials. The key needs to be provided as a realm key.", ProviderConfigProperty.STRING_TYPE, null),
+    // keyId header can be confusing if there is any key rotation, as key ids have to be immutable. It can lead
```
Resolved with [39100ed](https://github.com/keycloak/keycloak/commit/39100edc1281422399b045012acb3cb27062fbf6)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/model/Proof.java:
```diff
@@ -29,10 +31,19 @@
 
-    private Object proofObject;
+    @JsonProperty("jwt")
+    private String jwt;
```
And how do we deal with serialization?

francis-pouatcha replied:
Will be handled with issue https://github.com/keycloak/keycloak/issues/31363 as this is not a blocker.

francis-pouatcha commented on services/src/test/java/org/keycloak/protocol/oid4vc/model/ProofSerializationTest.java:
```diff
@@ -0,0 +1,16 @@
+        ObjectMapper objectMapper = new ObjectMapper();
+        String proofStr = " { \"proof_type\": \"jwt\", \"jwt\": \"ewogICJhbGciOiAiRVMyNTYiLAogICJ0eXAiOiAib3BlbmlkNHZjaS1wcm9vZitqd3QiLAogICJqd2siOiB7CiAgICAia3R5IjogIkVDIiwKICAgICJjcnYiOiAiUC0yNTYiLAogICAgIngiOiAiWEdkNU9GU1pwc080VkRRTUZrR3Z0TDVHU2FXWWE3SzBrNGhxUUdLbFBjWSIsCiAgICAieSI6ICJiSXFDaGhoVDdfdnYtYmhuRmVuREljVzVSUjRKTS1nME5sUi1qZGlHemNFIgogIH0KfQo.ewogICJpc3MiOiAib2lkNHZjaS1jbGllbnQiLAogI ...
+        Proof proof = objectMapper.readValue(proofStr, Proof.class);
+        assertEquals(ProofType.JWT, proof.getProofType());
```
What is the difference? Same Enum value sharing same value.

francis-pouatcha replied:
Will be handled with issue https://github.com/keycloak/keycloak/issues/31363 as this is not a blocker.

francis-pouatcha committed:
Processing more review comments

Signed-off-by: Francis Pouatcha <francis.pouatcha@adorsys.com>

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/SdJwtSigningService.java:
```diff
@@ -141,4 +178,164 @@ public String signCredential(VerifiableCredential verifiableCredential) {
+
+        Optional<Proof> optionalProof = getProofFromContext(vcIssuanceContext);
+
+        if (!optionalProof.isPresent()) {
```
Resolved with [3753e0f](https://github.com/keycloak/keycloak/pull/30692/commits/3753e0f914ae216c1ef7a2b87819617418f5e632)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/SdJwtSigningService.java:
```diff
@@ -141,4 +178,164 @@ public String signCredential(VerifiableCredential verifiableCredential) {
+        if(signatureVerifierContext==null){
+            throw new VCIssuerException("No verifier configured for " +jwsHeader.getAlgorithm());
+        }
+        if (!signatureVerifierContext.verify(jwsInput.getEncodedSignatureInput().getBytes("UTF-8"), jwsInput.getSignature())) {
```
Resolved with [3753e0f](https://github.com/keycloak/keycloak/pull/30692/commits/3753e0f914ae216c1ef7a2b87819617418f5e632)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/SdJwtSigningService.java:
```diff
@@ -141,4 +178,164 @@ public String signCredential(VerifiableCredential verifiableCredential) {
+        return Optional.ofNullable(vcIssuanceContext.getCredentialConfig())
+                .map(config -> config.getProofTypesSupported())
+                .flatMap(proofTypesSupported -> {
+                    if (proofTypesSupported == null) {
```
Resolved with [3753e0f](https://github.com/keycloak/keycloak/pull/30692/commits/3753e0f914ae216c1ef7a2b87819617418f5e632)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/SdJwtSigningService.java:
```diff
@@ -141,4 +178,164 @@ public String signCredential(VerifiableCredential verifiableCredential) {
+                        return Optional.empty();
+                    }
+
+                    ProofTypeJWT jwt = Optional.ofNullable(proofTypesSupported.getJwt())
```
Resolved with [3753e0f](https://github.com/keycloak/keycloak/pull/30692/commits/3753e0f914ae216c1ef7a2b87819617418f5e632)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/SdJwtSigningService.java:
```diff
@@ -141,4 +178,164 @@ public String signCredential(VerifiableCredential verifiableCredential) {
+        // As we limit accepted algorithm to the ones listed by the server, we can omit checking for "none"
+        // The Algorithm enum class does not list the none value anyway.
+        Optional.ofNullable(vcIssuanceContext.getCredentialConfig())
+                .map(config -> config.getProofTypesSupported())
```
Resolved with [3753e0f](https://github.com/keycloak/keycloak/pull/30692/commits/3753e0f914ae216c1ef7a2b87819617418f5e632)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/SdJwtSigningService.java:
```diff
@@ -141,4 +178,164 @@ public String signCredential(VerifiableCredential verifiableCredential) {
+        // The Algorithm enum class does not list the none value anyway.
+        Optional.ofNullable(vcIssuanceContext.getCredentialConfig())
+                .map(config -> config.getProofTypesSupported())
+                .map(proofTypesSupported -> proofTypesSupported.getJwt())
```
Resolved with [3753e0f](https://github.com/keycloak/keycloak/pull/30692/commits/3753e0f914ae216c1ef7a2b87819617418f5e632)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/SdJwtSigningService.java:
```diff
@@ -141,4 +178,164 @@ public String signCredential(VerifiableCredential verifiableCredential) {
+        Optional.ofNullable(vcIssuanceContext.getCredentialConfig())
+                .map(config -> config.getProofTypesSupported())
+                .map(proofTypesSupported -> proofTypesSupported.getJwt())
+                .map(jwt -> jwt.getProofSigningAlgValuesSupported())
```
Resolved with [3753e0f](https://github.com/keycloak/keycloak/pull/30692/commits/3753e0f914ae216c1ef7a2b87819617418f5e632)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/SdJwtSigningService.java:
```diff
@@ -141,4 +178,164 @@ public String signCredential(VerifiableCredential verifiableCredential) {
+        validateProofPayload(vcIssuanceContext, proofPayload);
+
+        SignatureVerifierContext signatureVerifierContext = getVerifier(jwk, jwsHeader.getAlgorithm().name());
+        if(signatureVerifierContext==null){
```
Resolved with [3753e0f](https://github.com/keycloak/keycloak/pull/30692/commits/3753e0f914ae216c1ef7a2b87819617418f5e632)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/model/ProofTypeDeserializer.java:
```diff
@@ -0,0 +1,34 @@
+package org.keycloak.protocol.oid4vc.model;
```
Resolved with [3753e0f](https://github.com/keycloak/keycloak/pull/30692/commits/3753e0f914ae216c1ef7a2b87819617418f5e632)

francis-pouatcha commented on services/src/test/java/org/keycloak/protocol/oid4vc/model/ProofSerializationTest.java:
```diff
@@ -0,0 +1,16 @@
+package org.keycloak.protocol.oid4vc.model;
```
Resolved with [3753e0f](https://github.com/keycloak/keycloak/pull/30692/commits/3753e0f914ae216c1ef7a2b87819617418f5e632)

francis-pouatcha commented on services/src/test/java/org/keycloak/protocol/oid4vc/model/ProofSerializationTest.java:
```diff
@@ -0,0 +1,16 @@
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.junit.Test;
+import static org.junit.Assert.*;
```
Resolved with [3753e0f](https://github.com/keycloak/keycloak/pull/30692/commits/3753e0f914ae216c1ef7a2b87819617418f5e632)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/model/ProofTypeSerializer.java:
```diff
@@ -0,0 +1,18 @@
+package org.keycloak.protocol.oid4vc.model;
```
Resolved with [3753e0f](https://github.com/keycloak/keycloak/pull/30692/commits/3753e0f914ae216c1ef7a2b87819617418f5e632)

francis-pouatcha commented:
@tnorimat both non addressed comment above will be handled with issue https://github.com/keycloak/keycloak/issues/31363 as this are not blockers.

@mposolda we shall be merging and closing this issue to avoid expensive merging activities.

francis-pouatcha committed:
Merged from upstream

Signed-off-by: Francis Pouatcha <francis.pouatcha@adorsys.com>

mposolda commented:
@francis-pouatcha I hope to merge if @wistefan and @tnorimat approve this PR to make sure that they are fine with latest changes.
@wistefan @tnorimat Could you please re-review this PR and possibly approve if it looks ok to you?

tnorimat commented:
@mposolda @francis-pouatcha Hello, I will re-review the PR.

francis-pouatcha committed:
Merged conficts from main

tnorimat commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/issuance/signing/OID4VCSdJwtIssuingEndpointTest.java:
```diff
@@ -0,0 +1,397 @@
+                JsonSerialization.mapper,
+                getKeyFromSession(session).getKid(),
+                Algorithm.ES256,
+                Format.SD_JWT_VC.toString(),
```
It might be `Format.SD_JWT_VC`

francis-pouatcha replied:
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

tnorimat commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/issuance/signing/OID4VCSdJwtIssuingEndpointTest.java:
```diff
@@ -0,0 +1,397 @@
+                JsonSerialization.mapper,
+                getKeyFromSession(session).getKid(),
+                Algorithm.ES256,
+                Format.SD_JWT_VC.toString(),
```
It might be `Format.SD_JWT_VC`

francis-pouatcha replied:
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

tnorimat commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/issuance/signing/OID4VCSdJwtIssuingEndpointTest.java:
```diff
@@ -0,0 +1,397 @@
+        componentExportRepresentation.setConfig(new MultivaluedHashMap<>(
+                Map.of(
+                        "algorithmType", List.of("ES256"),
+                        "tokenType", List.of(Format.SD_JWT_VC.toString()),
```
It might be `Format.SD_JWT_VC`

francis-pouatcha replied:
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

tnorimat commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/issuance/signing/OID4VCSdJwtIssuingEndpointTest.java:
```diff
@@ -0,0 +1,397 @@
+        ComponentExportRepresentation componentExportRepresentation = new ComponentExportRepresentation();
+        componentExportRepresentation.setName("sd-jwt-signing_identity_credential");
+        componentExportRepresentation.setId(UUID.randomUUID().toString());
+        componentExportRepresentation.setProviderId(Format.SD_JWT_VC.toString());
```
It might be `Format.SD_JWT_VC`

tnorimat commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/issuance/signing/OID4VCSdJwtIssuingEndpointTest.java:
```diff
@@ -0,0 +1,397 @@
+        ComponentExportRepresentation componentExportRepresentation = new ComponentExportRepresentation();
+        componentExportRepresentation.setName("sd-jwt-signing_test-credential");
+        componentExportRepresentation.setId(UUID.randomUUID().toString());
+        componentExportRepresentation.setProviderId(Format.SD_JWT_VC.toString());
```
It might be `Format.SD_JWT_VC`

francis-pouatcha replied:
Resolved with [d8cd042](https://github.com/keycloak/keycloak/pull/30692/commits/d8cd042dc629ebb73c665be5831483de8b9c2174)

tnorimat commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/issuance/signing/OID4VCSdJwtIssuingEndpointTest.java:
```diff
@@ -0,0 +1,397 @@
+        componentExportRepresentation.setConfig(new MultivaluedHashMap<>(
+                Map.of(
+                        "algorithmType", List.of("ES256"),
+                        "tokenType", List.of(Format.SD_JWT_VC.toString()),
```
It might be `Format.SD_JWT_VC`

francis-pouatcha replied:
Resolved with [d8cd042](https://github.com/keycloak/keycloak/pull/30692/commits/d8cd042dc629ebb73c665be5831483de8b9c2174)

tnorimat commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/issuance/signing/OID4VCSdJwtIssuingEndpointTest.java:
```diff
@@ -0,0 +1,397 @@
+        clientRepresentation.setEnabled(true);
+        Map<String, String> testCredentialAttributes = Map.of(
+                "vc.test-credential.expiry_in_s", "1800",
+                "vc.test-credential.format", Format.SD_JWT_VC.toString(),
```
It might be `Format.SD_JWT_VC`

francis-pouatcha replied:
Resolved with [d8cd042](https://github.com/keycloak/keycloak/pull/30692/commits/d8cd042dc629ebb73c665be5831483de8b9c2174)

tnorimat commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/issuance/signing/OID4VCSdJwtIssuingEndpointTest.java:
```diff
@@ -0,0 +1,397 @@
+        );
+        Map<String, String> identityCredentialAttributes = Map.of(
+                "vc.IdentityCredential.expiry_in_s", "31536000",
+                "vc.IdentityCredential.format", Format.SD_JWT_VC.toString(),
```
It might be `Format.SD_JWT_VC`

francis-pouatcha replied:
Resolved with [d8cd042](https://github.com/keycloak/keycloak/pull/30692/commits/d8cd042dc629ebb73c665be5831483de8b9c2174)

tnorimat reviewed (changes requested):
@francis-pouatcha Thank you. I added review commnets. Could you check them?

francis-pouatcha committed:
Removed unnecessary toString. Proofreading from review comments.

Signed-off-by: Francis Pouatcha <francis.pouatcha@adorsys.com>

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/OID4VCLoginProtocolFactory.java:
Resolved with [d8cd042](https://github.com/keycloak/keycloak/pull/30692/commits/d8cd042dc629ebb73c665be5831483de8b9c2174)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java:
```diff
@@ -301,39 +306,90 @@ public Response requestCredential(
-            throw new BadRequestException(getErrorResponse(ErrorType.UNSUPPORTED_CREDENTIAL_FORMAT));
+
+        // Check if at least one of both is available.
+        if(requestedCredentialId==null && requestedFormat==null){
```
Resolved with [d8cd042](https://github.com/keycloak/keycloak/pull/30692/commits/d8cd042dc629ebb73c665be5831483de8b9c2174)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java:
```diff
@@ -301,39 +306,90 @@ public Response requestCredential(
+        // Both Format and identifier are optional.
+        // If the credential_identifier is present, Format can't be present. But this implementation will
+        // tolerate the presence of both, waiting for clarity in specifications.
+        // This implementation will priviledge the presence of the credential config identifier.
```
Resolved with [d8cd042](https://github.com/keycloak/keycloak/pull/30692/commits/d8cd042dc629ebb73c665be5831483de8b9c2174)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java:
```diff
@@ -301,39 +306,90 @@ public Response requestCredential(
+
+        // Check if at least one of both is available.
+        if(requestedCredentialId==null && requestedFormat==null){
+            LOGGER.debugf("Missing both  configuration id and requested format. At least one shall be specified.");
```
Resolved with [d8cd042](https://github.com/keycloak/keycloak/pull/30692/commits/d8cd042dc629ebb73c665be5831483de8b9c2174)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java:
```diff
@@ -301,39 +306,90 @@ public Response requestCredential(
+
+        // resolve from identifier first
+        SupportedCredentialConfiguration supportedCredentialConfiguration = null;
+        if (requestedCredentialId!=null) {
```
Resolved with [d8cd042](https://github.com/keycloak/keycloak/pull/30692/commits/d8cd042dc629ebb73c665be5831483de8b9c2174)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java:
```diff
@@ -301,39 +306,90 @@ public Response requestCredential(
+
+        List<SupportedCredentialConfiguration> matchingConfigs;
+
+        if(Objects.equals(requestedFormat, Format.JWT_VC) || Objects.equals(requestedFormat, LDP_VC)) {
```
> How about importing JWT_VC the same as LDP_VC and SD_JWT_VC?
> import static org.keycloak.protocol.oid4vc.model.Format.JWT_VC;

Thanks

> I am not 100% sure but it might be replaced as follows:
>   List.of(SD_JWT_VC, LDP_VC, JWT_VC).contains(requestedFormat)

Wouldn't get the purpose. As i am resolving case per case.

> I think that SD_JWT_VC is more plausible than other two VC format, to how about the following?
> if (Objects.equals(requestedFormat, SD_JWT_VC) || Objects.equals(requestedFormat, LDP_VC) || Objects.equals(requestedFormat, JWT_VC)) {

I don't need to check all of them in one shot, but will start with JWT_VC

francis-pouatcha replied:
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java:
```diff
@@ -301,39 +306,90 @@ public Response requestCredential(
+            throw new BadRequestException(getErrorResponse(ErrorType.MISSING_CREDENTIAL_CONFIG));
+        }
+
+        return matchingConfigs.iterator().next();
```
For now we a taking the first of the best matches match. No sure if we will ever have more than on matching config here. I don't see the use case for the moment.

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java:
```diff
@@ -301,39 +306,90 @@ public Response requestCredential(
+        if(Objects.equals(requestedFormat, Format.JWT_VC) || Objects.equals(requestedFormat, LDP_VC) || Objects.equals(requestedFormat, SD_JWT_VC)) {
+            responseVO.setCredential(theCredential);
+        } else {
+            throw new BadRequestException(getErrorResponse(ErrorType.UNSUPPORTED_CREDENTIAL_TYPE));
```
Went for your first suggestion with check of set membership.

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java:
```diff
@@ -371,14 +421,14 @@ private AuthenticationManager.AuthResult getAuthResult(WebApplicationException e
+    private Object getCredential(AuthenticationManager.AuthResult authResult, SupportedCredentialConfiguration credentialConfig, CredentialRequest credentialRequestVO) {
 
-        List<OID4VCClient> clients = getClientsOfType(vcType, format);
+        List<OID4VCClient> clients = getClientsOfScope(credentialConfig.getScope(), credentialConfig.getFormat());
```
No sure why @wistefan proceeded like that. I will bring this in a proper issue: https://github.com/keycloak/keycloak/issues/31493

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/VCSigningServiceProviderFactory.java:
```diff
@@ -44,13 +44,15 @@ public interface VCSigningServiceProviderFactory extends ComponentFactory<Verifi
+                .checkRequired(SigningProperties.ALGORITHM_TYPE.asConfigProperty());
         validateSpecificConfiguration(session, realm, model);
     }
 
```
It is no longer there. Validate varies by format.

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/JwtProofBasedSigningService.java:
```diff
@@ -0,0 +1,208 @@
+        //        Optional.ofNullable(proofPayload.getIssuer())
+        //                .filter(proofIssuer -> Objects.equals(azp, proofIssuer))
+        //                .orElseThrow(() -> new VCIssuerException("Issuer claim must be null for preauthorized code else the clientId of the client making the request: " + azp));
+
```
> Could you confirm the following?
> 
> A client (1) that sends a token request and receives an access token is the one whose protocol is oidc.

Yes. To obtain a token, i need an oidc client. Or do you think i can use an oid4vci client to obtain a token?

> A client (2) that sends a credential request and receives a VC is the one whose protocol is oid4vci.

Yes. Unless we have a way of issuing a token with an oidc client.

> Both shold be the same, but Keycloak does not do that ( azp in an access token shows the client (1) while a sender of Key binding proof is the client (2)).

This is why i did not check the azp here and left it commented. We need a concept.

> To get around it, the client (1)'s access token includes audiences that includes the client (2), and the method check it.

Ok. I will try this. As client access token can have many audiences. Thank you.

> If those above are right, IMO, it might be preferable to require only one client for OID4VCI.
> 
> For example, [POTENTIAL Interop Event / Track 2 / Full Profile](https://www.authlete.com/developers/oid4vci/#45-potential-interop-event--track-2--full-profile) requires DPoP Proof JWT on Token Endpiont and Credential Endpoint, which means that a sender of a token request should be the sender of a credential request. I am afraid that the current Keycloak's treatment of clients for OID4VCI does not realize such the use-case.

Correct. I will try the approach with many audiences.

> 
> Also, for tracking of VC issuance for audit, it might be preferable that one client go through all phases from getting a credential offer, an access token, and a VC.

Yes.

I will create a separate ticket to implement this: https://github.com/keycloak/keycloak/issues/31496

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/mappers/OID4VCMapper.java:
```diff
@@ -97,15 +97,15 @@ public void close() {
-        var optionalTypes = Optional.ofNullable(mapperModel.getConfig().get(SUPPORTED_CREDENTIALS_KEY));
-        if (optionalTypes.isEmpty()) {
+    public boolean isScopeSupported(String credentialScope) {
+        var optionalScopes = Optional.ofNullable(mapperModel.getConfig().get(SUPPORTED_CREDENTIALS_KEY));
```
Line 44 is used externally in the client config. I` will try synchronize this with @wistefan . I also still confused with the separation between credentialType and scope.

Moved this to ticket: https://github.com/keycloak/keycloak/issues/31497

francis-pouatcha committed:
Processed review comments, formating, coding rules

Signed-off-by: Francis Pouatcha <francis.pouatcha@adorsys.com>

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java:
```diff
@@ -301,39 +306,90 @@ public Response requestCredential(
+        SupportedCredentialConfiguration supportedCredentialConfiguration = null;
+        if (requestedCredentialId!=null) {
+            supportedCredentialConfiguration = supportedCredentials.get(requestedCredentialId);
+            if(supportedCredentialConfiguration==null){
```
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java:
```diff
@@ -301,39 +306,90 @@ public Response requestCredential(
+            }
+            // Then for format. We know spec does not allow both parameter. But we are tolerant if you send both
+            // Was found by id, check that the format matches.
+            if (requestedFormat!=null && !requestedFormat.equals(supportedCredentialConfiguration.getFormat())){
```
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java:
```diff
@@ -301,39 +306,90 @@ public Response requestCredential(
+            }
+        }
+
+        if(supportedCredentialConfiguration==null && requestedFormat!=null) {
```
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java:
```diff
@@ -301,39 +306,90 @@ public Response requestCredential(
+        if(supportedCredentialConfiguration==null && requestedFormat!=null) {
+            // Search by format
+            supportedCredentialConfiguration = getSupportedCredentialConfiguration(credentialRequestVO, supportedCredentials, requestedFormat);
+            if(supportedCredentialConfiguration==null) {
```
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java:
```diff
@@ -301,39 +306,90 @@ public Response requestCredential(
+    private SupportedCredentialConfiguration getSupportedCredentialConfiguration(CredentialRequest credentialRequestVO, Map<String, SupportedCredentialConfiguration> supportedCredentials, String requestedFormat) {
+        // 1. Format resolver
+        List<SupportedCredentialConfiguration> configs = supportedCredentials.values().stream()
+                .filter(supportedCredential -> Objects.equals(supportedCredential.getFormat(),requestedFormat))
```
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java:
```diff
@@ -301,39 +306,90 @@ public Response requestCredential(
-            default -> throw new BadRequestException(
-                    getErrorResponse(ErrorType.UNSUPPORTED_CREDENTIAL_TYPE));
+        Object theCredential = getCredential(authResult, supportedCredentialConfiguration, credentialRequestVO);
+        if(Objects.equals(requestedFormat, Format.JWT_VC) || Objects.equals(requestedFormat, LDP_VC) || Objects.equals(requestedFormat, SD_JWT_VC)) {
```
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java:
```diff
@@ -301,39 +306,90 @@ public Response requestCredential(
+                    .collect(Collectors.toList());
+        } else {
+            throw new BadRequestException(getErrorResponse(ErrorType.UNSUPPORTED_CREDENTIAL_FORMAT));
+        }
```
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java:
```diff
@@ -301,39 +306,90 @@ public Response requestCredential(
+            matchingConfigs = configs.stream()
+                    .filter(supportedCredential -> Objects.equals(supportedCredential.getCredentialDefinition(), credentialRequestVO.getCredentialDefinition()))
+                    .collect(Collectors.toList());
+        } else if (Objects.equals(requestedFormat, Format.SD_JWT_VC)) {
```
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java:
```diff
@@ -371,14 +421,14 @@ private AuthenticationManager.AuthResult getAuthResult(WebApplicationException e
-     * @param format           format of the credential to be created
+     * @param authResult    authResult containing the userSession to create the credential for
+     * @param credentialConfig    the supported credential configuration
+     * @param credentialRequestVO
```
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java:
```diff
@@ -427,19 +491,20 @@ private Response getErrorResponse(ErrorType errorType) {
-    private List<OID4VCClient> getClientsOfType(String vcType, String format) {
-        LOGGER.debugf("Retrieve all clients of type %s, supporting format %s", vcType, format.toString());
+    // Return all {@link  OID4VCClient}s that support the given scope and format
+    // Scope might be different from vct. In the case of sd-jwt for eaxample
```
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/model/CredentialConfigId.java:
```diff
@@ -0,0 +1,37 @@
+    private String value;
+
+    public static CredentialConfigId from(String value) {
+
```
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/model/CredentialConfigId.java:
```diff
@@ -0,0 +1,37 @@
+
+    public static CredentialConfigId from(String value) {
+
+        return value==null ? null : new CredentialConfigId(value);
```
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/model/SupportedCredentialConfiguration.java:
```diff
@@ -89,6 +94,30 @@ public String getFormat() {
+     * @return
+     */
+    public VerifiableCredentialType deriveType() {
+        if(Objects.equals(format, Format.SD_JWT_VC)) {
```
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/model/SupportedCredentialConfiguration.java:
```diff
@@ -169,6 +198,15 @@ public SupportedCredentialConfiguration setVct(String vct) {
+
     public ProofTypesSupported getProofTypesSupported() {
         return proofTypesSupported;
     }
```
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/model/SupportedCredentialConfiguration.java:
```diff
@@ -223,6 +262,9 @@ public static SupportedCredentialConfiguration fromDotNotation(String credential
+                .ifPresent(supportedCredentialConfiguration::setCredentialDefinition);
 
         String displayKeyPrefix = credentialId + DOT_SEPARATOR + DISPLAY_KEY + DOT_SEPARATOR;
         List<DisplayObject> displayList = dotNotated.entrySet().stream()
```
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/model/VerifiableCredentialType.java:
```diff
@@ -0,0 +1,35 @@
+    private String value;
+
+    public static VerifiableCredentialType from(String value){
+        return value==null? null : new VerifiableCredentialType(value);
```
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/SigningService.java:
```diff
@@ -17,9 +17,16 @@
+import org.keycloak.jose.jwk.JWKParser;
+import org.keycloak.jose.jwk.OKPPublicJWK;
 import org.keycloak.models.KeycloakSession;
+import org.keycloak.protocol.oid4vc.model.Format;
```
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/SigningService.java:
```diff
@@ -34,18 +41,66 @@ public abstract class SigningService<T> implements VerifiableCredentialsSigningS
 
+    @Override
+    public String locator() {
+        // Future implementation might consider credential type or even cofiguration specific signers.
```
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/SigningService.java:
```diff
@@ -34,18 +41,66 @@ public abstract class SigningService<T> implements VerifiableCredentialsSigningS
+        return signatureProvider.verifier(getKeyWraper(jwk, jwsAlgorithm, KeyUse.SIG));
+    }
+
+    private KeyWrapper getKeyWraper(JWK jwk, String algorithm, KeyUse keyUse) {
```
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/VerifiableCredentialsSigningService.java:
```diff
@@ -30,8 +33,32 @@ public interface VerifiableCredentialsSigningService<T> extends Provider {
+        return (format==null? "" : format) + LOCATION_SEPARATOR +
+                (credentialType==null?"":credentialType.getValue()) + LOCATION_SEPARATOR +
+                (vcConfigId==null?"":vcConfigId.getValue());
+    }
```
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/VCSigningServiceProviderFactory.java:
```diff
@@ -44,13 +44,15 @@ public interface VCSigningServiceProviderFactory extends ComponentFactory<Verifi
 
```
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/VCSigningServiceProviderFactory.java:
```diff
@@ -44,13 +44,15 @@ public interface VCSigningServiceProviderFactory extends ComponentFactory<Verifi
 
     public static ProviderConfigurationBuilder configurationBuilder() {
```
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/SdJwtSigningServiceProviderFactory.java:
```diff
@@ -50,6 +51,10 @@ public VerifiableCredentialsSigningService create(KeycloakSession session, Compo
+        String vcConfigId = model.get(SigningProperties.VC_CONFIG_ID.getKey());
 
         List<String> visibleClaims = Optional.ofNullable(model.get(SigningProperties.VISIBLE_CLAIMS.getKey()))
                 .map(visibileClaims -> visibileClaims.split(","))
```
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/SdJwtSigningServiceProviderFactory.java:
```diff
@@ -89,11 +97,15 @@ public String getId() {
                 .checkRequired(SigningProperties.TOKEN_TYPE.asConfigProperty())
                 .checkInt(SigningProperties.DECOYS.asConfigProperty(), true);
+        // Make sure VCT is set if vc config id is set.
+        if(model.get(SigningProperties.VC_CONFIG_ID.getKey())!=null){
```
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/JwtProofBasedSigningService.java:
```diff
@@ -0,0 +1,208 @@
+                        "Proof not produced for this audience. Audience claim must be: " + credentialIssuer + " but are " + Arrays.asList(proofPayload.getAudience())));
+
+        // Validate mandatory iat.
+        // I do not understand the rationale behind requiring a issue time if we are not checking expiration.
```
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/JwtProofBasedSigningService.java:
```diff
@@ -0,0 +1,208 @@
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.protocol.oid4vc.issuance.VCIssuanceContext;
+import org.keycloak.protocol.oid4vc.issuance.VCIssuerException;
+import org.keycloak.protocol.oid4vc.model.Format;
```
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/JwtProofBasedSigningService.java:
```diff
@@ -0,0 +1,208 @@
+
+        SignatureVerifierContext signatureVerifierContext = getVerifier(jwk, jwsHeader.getAlgorithm().name());
+        if (signatureVerifierContext == null) {
+            throw new VCIssuerException("No verifier configured for " +jwsHeader.getAlgorithm());
```
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/SdJwtSigningService.java:
```diff
@@ -46,53 +54,79 @@
         if (signingKey == null) {
             throw new SigningServiceException(String.format("No key for id %s and algorithm %s available.", keyId, algorithmType));
         }
+        // @Francis: keyId header can be confusing if there is any key rotation, as key ids have to be immutable. It can lead
```
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

francis-pouatcha commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/issuance/signing/SdJwtSigningServiceTest.java:
```diff
@@ -54,7 +59,6 @@
-    private static ObjectMapper OBJECT_MAPPER = new ObjectMapper();
     private static KeyWrapper rsaKey = getRsaKey();
 
     // If an unsupported algorithm is provided, the JWT Sigining Service should not be instantiated.
```
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

francis-pouatcha commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/issuance/signing/SdJwtSigningServiceTest.java:
```diff
@@ -74,8 +78,9 @@ public void testUnsupportedAlgorithm() throws Throwable {
+                                    CredentialConfigId.from("test-credential")));
         } catch (RunOnServerException ros) {
             throw ros.getCause();
         }
```
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

francis-pouatcha commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/issuance/signing/OID4VCTest.java:
```diff
@@ -353,4 +352,18 @@ public long currentTimeMillis() {
         }
     }
 
+    protected ProtocolMapperRepresentation getUserAttributeMapper(String subjectProperty, String atributeName, String supportedCredentialTypes) {
```
Resolved with [5f99e16](https://github.com/keycloak/keycloak/pull/30692/commits/5f99e1671d30c85c0718a56dd88d3508f4a20d09)

francis-pouatcha committed:
Merge branch 'keycloak:main' into issue-30525

francis-pouatcha committed:
Review comments. Final fields on immutable classes.

Signed-off-by: Francis Pouatcha <francis.pouatcha@adorsys.com>

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/model/VerifiableCredentialType.java:
```diff
@@ -0,0 +1,35 @@
+ * @author <a href="mailto:francis.pouatcha@adorsys.com">Francis Pouatcha</a>
+ */
+public class VerifiableCredentialType {
+    private String value;
```
Resolved with [9717c87](https://github.com/keycloak/keycloak/pull/30692/commits/9717c876021ed88f5b6b1341ba5f07e9ff0b67c2)

francis-pouatcha commented on services/src/main/java/org/keycloak/protocol/oid4vc/model/CredentialConfigId.java:
```diff
@@ -0,0 +1,37 @@
+ * @author <a href="mailto:francis.pouatcha@adorsys.com">Francis Pouatcha</a>
+ */
+public class CredentialConfigId {
+    private String value;
```
Resolved with [9717c87](https://github.com/keycloak/keycloak/pull/30692/commits/9717c876021ed88f5b6b1341ba5f07e9ff0b67c2)

francis-pouatcha commented:
@tnorimat i shall have all review comment addressed.

tnorimat reviewed (approved):
LGTM

tnorimat commented:
@francis-pouatcha Thank you.
@wistefan  @mposolda I approved the PR. If possible, could you check the PR?

mposolda reviewed (approved):
@francis-pouatcha @tnorimat Thanks for the participation and review of this PR!
